[GUI] Add address-label to sendconfirm-popup

Mrs-X opened the pull request:
Adds the label of the recipient to the confirmation-popup when sending funds because the label is much easier to verify than the address itself.

It's intentionally NOT added to the list of recipients if you send to more than 1 recipient because the space on that list is already very crowded and would make it even less useful.

Screenshot:
![Sendpopup](https://user-images.githubusercontent.com/22873440/73354953-439bcb80-4297-11ea-9c8a-c8da1b676ec8.png)

furszy commented:
One thing, at this point the address label was not stored in the addressbook yet. So the user cannot verify the recipient only checking it. It can be a duplicated label.

Example: you paste a payment request in the send screen, that will autocomplete every edit texts. The address, amount and label. Imagine that the URI has a different address and the same label than one of your own addresses. The confirmation dialog will appear with the same addressbook stored name but with a different address.

Saying that, I'm just mentioning it to not end up with users verifying only the label as it could be misleading. But it's not something for this PR, another PR work.

--------------------

Side from that, I think that this indeed a good add for a single address send.

Graphically: would invert the edit texts order, first label, then address below. Decrease the vertical spacing between both fields, make the address size 2px smaller than the label and center everything vertically.


```
Sending to:           Budget-Payment
                       < address >
```

random-zebra commented on src/qt/pivx/sendconfirmdialog.cpp:
```diff
@@ -118,18 +118,22 @@ void TxDetailDialog::setData(WalletModel *model, const QModelIndex &index){
-    if(tx.getRecipients().size() == 1){
+    if (tx.getRecipients().size() == 1) {
         ui->textSend->setText(tx.getRecipients().at(0).address);
+        ui->textSendLabel->setText(tx.getRecipients().at(0).label);
```
performance nit: wouldn't it be better to save the `QList` rather than making three separate calls to `tx.getRecipients()` ?

Mrs-X replied:
Not sure, you'd have to  additionally create a SendCoinsRecipient object and an int.
But I've changed it anyway...

random-zebra commented on src/qt/pivx/sendconfirmdialog.cpp:
```diff
@@ -118,18 +118,22 @@ void TxDetailDialog::setData(WalletModel *model, const QModelIndex &index){
         ui->pushOutputs->setVisible(false);
-    }else{
+    }
+    else {
```
styling nit: I think the (most widely used) convention in the codebase is to put the else on the same line of the closing bracket (with spaces) `} else {`

Mrs-X replied:
Automatic formatting of IDE adjusted

Mrs-X commented:
All nits addressed, segfault (hopefully, because I can't reproduce it) fixed and layout changed as proposed by @furszy : 
![Send_Confirm2](https://user-images.githubusercontent.com/22873440/73545576-dd9c7900-443b-11ea-9a71-1941ca41f989.png)

furszy commented:
Minor thing that just saw on the image: the "Total Amount" is missing the colon at the end like the rest of the titles.

Mrs-X commented:
My findings about this segfault:
- hiding the widget which contains the address (and a placeholder) is according to the Qt documentation perfectly fine, so I assume that this change is just an indicator of some other problem
- nonetheless I have removed that line of code, the only difference is that we have now one empty line below the text "Sending to n recipients" when sending to more than one recipient, something I can live with.

I've also changed "Total amount" to "Total amount:"

furszy commented:
Hmm, this is affecting the transaction detail too. Let's correct it instead of have the extra spacing. Will affect both scenarios, send confirmation and transaction detail.

And well, most importantly.. After all of the work done to have the UI looking good, we really need to try as much as possible to not add bad UI looking things. Even small things, that is how the old UI ended up being an awful monster.

-------------------------------

I don't know where was exactly zebra's issue (as the `labelOutputs` object that he mentioned above it's not on the layout anymore). -- would be good to have more information.

But I see a problem on how the layout was arranged, the `textSend` has to be inside `contentOutputsLabel` and not on its own extra widget below. That will let you hide the label in case of the user don't having a set label for the address. And the bottom margin of the `contentOutputsLabel` needs to be added back in order to be inline with the rest of the rows spacing.


I have solved it here, code patch:

```
Index: src/qt/pivx/forms/sendconfirmdialog.ui
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/qt/pivx/forms/sendconfirmdialog.ui	(date 1581543049000)
+++ src/qt/pivx/forms/sendconfirmdialog.ui	(date 1581547860331)
@@ -219,7 +219,7 @@
               <x>0</x>
               <y>0</y>
               <width>556</width>
-              <height>703</height>
+              <height>690</height>
              </rect>
             </property>
             <property name="autoFillBackground">
@@ -336,11 +336,14 @@
                 </item>
                 <item>
                  <widget class="QWidget" name="contentOutputsLabel" native="true">
-                  <layout class="QHBoxLayout" name="horizontalLayout_11" stretch="0,0,0">
+                  <layout class="QHBoxLayout" name="horizontalLayout_11" stretch="0,1,0">
                    <property name="spacing">
                     <number>0</number>
                    </property>
                    <property name="leftMargin">
+                    <number>0</number>
+                   </property>
+                   <property name="topMargin">
                     <number>0</number>
                    </property>
                    <property name="rightMargin">
@@ -351,15 +354,9 @@
                    </property>
                    <item>
                     <widget class="QLabel" name="labelSend">
-                     <property name="minimumSize">
-                      <size>
-                       <width>120</width>
-                       <height>0</height>
-                      </size>
-                     </property>
                      <property name="maximumSize">
                       <size>
-                       <width>120</width>
+                       <width>16777215</width>
                        <height>16777215</height>
                       </size>
                      </property>
@@ -369,10 +366,23 @@
                     </widget>
                    </item>
                    <item>
-                    <widget class="QLabel" name="textSendLabel">
-                     <property name="text">
-                      <string>Name/Label</string>
-                     </property>
+                    <widget class="QWidget" name="widget" native="true">
+                     <layout class="QVBoxLayout" name="verticalLayout_6">
+                      <item>
+                       <widget class="QLabel" name="textSendLabel">
+                        <property name="text">
+                         <string>Name/Label</string>
+                        </property>
+                       </widget>
+                      </item>
+                      <item>
+                       <widget class="QLabel" name="textSend">
+                        <property name="text">
+                         <string>D7VFR83SQbiezrW72hjcWJtcfip5krte2Z</string>
+                        </property>
+                       </widget>
+                      </item>
+                     </layout>
                     </widget>
                    </item>
                    <item>
@@ -397,53 +407,6 @@
                      </property>
                      <property name="text">
                       <string/>
-                     </property>
-                    </widget>
-                   </item>
-                  </layout>
-                 </widget>
-                </item>
-                <item>
-                 <widget class="QWidget" name="contentOutputsAddress" native="true">
-                  <layout class="QHBoxLayout" name="horizontalLayout_outputs" stretch="0,0">
-                   <property name="spacing">
-                    <number>0</number>
-                   </property>
-                   <property name="leftMargin">
-                    <number>0</number>
-                   </property>
-                   <property name="topMargin">
-                    <number>0</number>
-                   </property>
-                   <property name="rightMargin">
-                    <number>12</number>
-                   </property>
-                   <property name="bottomMargin">
-                    <number>9</number>
-                   </property>
-                   <item>
-                    <widget class="QLabel" name="labelSendLabel">
-                     <property name="minimumSize">
-                      <size>
-                       <width>120</width>
-                       <height>0</height>
-                      </size>
-                     </property>
-                     <property name="maximumSize">
-                      <size>
-                       <width>120</width>
-                       <height>16777215</height>
-                      </size>
-                     </property>
-                     <property name="text">
-                      <string/>
-                     </property>
-                    </widget>
-                   </item>
-                   <item>
-                    <widget class="QLabel" name="textSend">
-                     <property name="text">
-                      <string>D7VFR83SQbiezrW72hjcWJtcfip5krte2Z</string>
                      </property>
                     </widget>
                    </item>
Index: src/qt/pivx/sendconfirmdialog.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/qt/pivx/sendconfirmdialog.cpp	(date 1581543049000)
+++ src/qt/pivx/sendconfirmdialog.cpp	(date 1581547433092)
@@ -96,10 +96,11 @@
         ui->textId->setText(hash.left(20) + "..." + hash.right(20));
         ui->textId->setTextInteractionFlags(Qt::TextSelectableByMouse);
         if (tx->vout.size() == 1) {
-            ui->textSend->setText(address);
+            ui->textSendLabel->setText(address);
         } else {
-            ui->textSend->setText(QString::number(tx->vout.size()) + " recipients");
+            ui->textSendLabel->setText(QString::number(tx->vout.size()) + " recipients");
         }
+        ui->textSend->setVisible(false);
 
         ui->textInputs->setText(QString::number(tx->vin.size()));
         ui->textConfirmations->setText(QString::number(rec->status.depth));
@@ -128,12 +129,17 @@
     int nRecipients = tx.getRecipients().size();
     if (nRecipients == 1) {
         SendCoinsRecipient recipient = tx.getRecipients().at(0);
-        ui->textSend->setText(recipient.address);
-        ui->textSendLabel->setText(recipient.label);
+        if (recipient.label.isEmpty()) { // If there is no label, then do not show the blank space.
+            ui->textSendLabel->setText(recipient.address);
+            ui->textSend->setVisible(false);
+        } else {
+            ui->textSend->setText(recipient.address);
+            ui->textSendLabel->setText(recipient.label);
+        }
         ui->pushOutputs->setVisible(false);
     } else {
         ui->textSendLabel->setText(QString::number(nRecipients) + " recipients");
-        ui->textSend->setText("");
+        ui->textSend->setVisible(false);
     }
     ui->textInputs->setText(QString::number(tx.getTransaction()->vin.size()));
     ui->textFee->setText(BitcoinUnits::formatWithUnit(nDisplayUnit, txFee, false, BitcoinUnits::separatorAlways));
```

random-zebra commented:
Fixed `textSend` label inside the widget and rearranged the dialog layout a bit:

![sendconfirm1](https://user-images.githubusercontent.com/18186894/76546834-73401500-648c-11ea-9837-5b408d4ee30c.png)

<br>

![sendconfirm2](https://user-images.githubusercontent.com/18186894/76546842-7a672300-648c-11ea-969b-5d0b2f0144b6.png)

random-zebra commented:
Rebased on master

random-zebra commented:
Superseded  by #1447 . Closing.
